### PR TITLE
[ROCm] Add partitioned buffer approach for scatter add op

### PR DESCRIPTION
**Pull request by jataylo**
It has been observed that in the case of heavy contended atomics poor performance is being achieved.

To solve this problem while minimizing kernel overhead this PR proposes an fx pass which will replace the index_put operation with an alternative scatter approach.

Algorithm:
1. Enumerate scatter operations: operation_id = [0, 1, 2, ..., N-1]
2. Assign to partitions: partition_id = operation_id % num_partitions
3. Create expanded buffers along scatter_dim: size = num_partitions × dim_size
4. Adjust indices: adjusted_idx = original_idx + (partition_id × dim_size)
5. Perform partitioned scatter with reduced contention
6. Reduce across partitions: sum(partitions, dim=scatter_dim)

This will reduce atomic contention at the cost of memory usage. In order to combat this we have built heuristics around the total number of partitions for the expanded buffer, as well as setting a cap on how large these expanded tensors can be (currently 10% of GPU memory)

Note the heuristic cannot be perfect as we do not know the true indices data at compile time, in real world models the indices will have duplicates and not be uniformly distributed which increases atomic contention, currently this cannot be modelled and we have to estimate contention based on input and output buffer sizes.

Benchmark code: https://gist.github.com/jataylo/dd3a6353ad2859efd65fa87b28aa3ebd
This code executes 3 index_add ops to 3 seperate buffers. 
N = 1000000
D = 100
n = 501

values = float32 [N,D]
indices = int64 [N]
output = float32 [n, D]

For each run we modify the range of randint to simulate various levels of atomic contention

Gathered two sets of results, one with partitioned_scatter_enabled=True, the other partitioned_scatter_enabled=False

**MI300**
uniform_range | no_compile_ms | compile_ms   (partitioned_scatter_enabled=False) | compile_ms   (partitioned_scatter_enabled=True) | speedup
-- | -- | -- | -- | --
0-0 | 85.52 | 28.50 | 3.55 | 8.03
0-1 | 46.99 | 15.66 | 2.47 | 6.33
0-3 | 25.16 | 8.31 | 2.20 | 3.78
0-7 | 12.92 | 4.32 | 1.63 | 2.66
0-15 | 6.66 | 4.24 | 1.60 | 2.66
0-31 | 3.43 | 3.19 | 1.33 | 2.40
0-63 | 1.79 | 1.62 | 1.32 | 1.23
0-127 | 1.76 | 1.59 | 1.24 | 1.28
0-255 | 1.73 | 1.32 | 1.24 | 1.07
0-500 | 1.61 | 1.27 | 1.23 | 1.04

**H100**
uniform_range | no_compile_ms | compile_ms     (partitioned_scatter_enabled=False) | compile_ms     (partitioned_scatter_enabled=True)
-- | -- | -- | --
0-0 | 19.842156 | 5.504691 | 0.756135
0-1 | 19.516249 | 5.526914 | 0.779779
0-3 | 10.450396 | 3.079321 | 0.702327
0-7 | 5.417728 | 1.906823 | 0.683553
0-15 | 3.545023 | 1.570733 | 0.655154
0-31 | 2.631531 | 1.223358 | 0.641384
0-63 | 2.629384 | 0.832651 | 0.63534
0-127 | 2.629735 | 0.726054 | 0.768287
0-255 | 2.62846 | 0.625181 | 1.317291
0-500 | 2.629526 | 0.548981 | 1.724292

We can see this could potentially benefit H100 on worst-case examples but would degrade perf in the best case, the atomic add cost on MI300 is heavier meaning this is more beneficial.

On MI300 we can see a mixed bag of e2e model improvements 
https://hud.pytorch.org/benchmark/v3/dashboard/compiler_inductor?renderGroupId=main&time.start=2025-11-05T00%3A00%3A00.000Z&time.end=2025-12-04T02%3A00%3A00.000Z&filters.repo=pytorch%2Fpytorch&filters.benchmarkName=compiler&filters.mode=training&filters.dtype=amp&filters.deviceName=rocm+%28mi300x%29&filters.device=rocm&filters.suite=all&filters.compiler=default&lcommit.commit=38c42c575d342a7ea6f4a555bf845071e03b5f35&lcommit.workflow_id=19635538449&lcommit.date=2025-11-24T14%3A00%3A00Z&lcommit.branch=refs%2Ftags%2Fciflow%2Finductor-perf-test-nightly-rocm-mi300%2F168073&rcommit.commit=fedb7f15d177a259bf25c94e888137e0a9a69a81&rcommit.workflow_id=19856622912&rcommit.date=2025-12-02T12%3A00%3A00Z&rcommit.branch=refs%2Ftags%2Fciflow%2Finductor-perf-test-nightly-rocm-mi300%2F168073&lbranch=refs%2Ftags%2Fciflow%2Finductor-perf-test-nightly-rocm-mi300%2F168073&rbranch=refs%2Ftags%2Fciflow%2Finductor-perf-test-nightly-rocm-mi300%2F168073&maxSampling=110

Due to mixed-bag of results we will initially enable this as non default feature but testing passed CI with this enabled here 
https://hud.pytorch.org/pytorch/pytorch/pull/168073?sha=fedb7f15d177a259bf25c94e888137e0a9a69a81

Note there are improvements to make after this lands:

1. Add dynamic shape support, needs to be conservative here to not explode memory usage.
2. Update IR and codegen directly to avoid iota op and needing to update indices via torch ops, we can likely do this in store codegen itself.
3. Develop new implementations for memory constrained environments

cc @jeffdaily @sunway513 @jithunnair-amd @pruthvistony @ROCmSupport @hongxiayang @naromero77amd @pragupta @jerrymannil @xinyazhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @kadeng @muchulee8 @amjames @chauhang @aakhundov @coconutruben @dllehr-amd @chenyang78

### CODE REVIEW

File: torch/_inductor/fx_passes/reduced_atomic_contention.py [hunk 1]
```diff
@@ -0,0 +1,582 @@
...
+    input_meta = _get_tensor_meta(input_node)
+    index_meta = _get_tensor_meta(index_node)
+    if not input_meta or not index_meta:
+        return False
+
+    # Skip unsupported cases
+    if isinstance(input_meta["numel"], torch.SymInt) or isinstance(
+        index_meta["numel"], torch.SymInt
+    ):
+        log.debug("Skipping: dynamic shapes not supported")
```
Comment (eellison) [inline torch/_inductor/fx_passes/reduced_atomic_contention.py:106]: what is preventing dynamic shapes?

File: torch/_inductor/fx_passes/reduced_atomic_contention.py [hunk 2]
```diff
@@ -0,0 +1,582 @@
...
+        return False
+
+    if input_meta["dtype"] == torch.bool or index_meta["dtype"] == torch.bool:
```
Comment (eellison) [inline torch/_inductor/fx_passes/reduced_atomic_contention.py:109]: similarly? whats preventing bool?

File: torch/_inductor/fx_passes/reduced_atomic_contention.py [hunk 3]
```diff
@@ -0,0 +1,582 @@
...
+    PatternMatcherPass,
+    register_graph_pattern,
+)
+
+
+log = logging.getLogger(__name__)
+aten = torch.ops.aten
+prims = torch.ops.prims
+
+# Constants for partition configuration
```
Comment (eellison) [inline torch/_inductor/fx_passes/reduced_atomic_contention.py]: worth config-ing ?

File: torch/_inductor/fx_passes/reduced_atomic_contention.py [hunk 4]
```diff
@@ -0,0 +1,582 @@
...
+        args=(values, flat_values_shape),
+    )
+    _set_fake_tensor_meta(
+        flat_values, flat_values_shape, values_meta.dtype, device, fake_mode
+    )
+
+    return flat_index, flat_values
+
+
+def _create_partitioned_scatter(
```
Comment (eellison) [inline torch/_inductor/fx_passes/reduced_atomic_contention.py]: I think this would be more concise with either register_replacement or, [register_graph_pattern + replace_by_example](https://github.com/pytorch/pytorch/blob/1e49fa9204393384e127b2895f7f30986bbe1215/torch/_inductor/fx_passes/post_grad.py#L1509-L1532)

File: torch/_inductor/fx_passes/reduced_atomic_contention.py [hunk 5]
```diff
@@ -0,0 +1,582 @@
...
+        CallFunction(aten.index_put_.default, Arg(), Arg(), Arg(), True),
+        extra_check=validate_match,
+        pass_dict=partitioned_scatter_patterns,  # type: ignore[arg-type]
+    )(create_replacement)
+
+
+def validate_match(match: Match) -> bool:
+    """Check if pattern match should be optimized."""
+    output_node = match.output_node()
+    if not output_node or not hasattr(output_node, "args") or len(output_node.args) < 4:
```
Comment (eellison) [inline torch/_inductor/fx_passes/reduced_atomic_contention.py:76]: See, something like https://github.com/pytorch/pytorch/blob/876deaef7d88790c5a92b7f3c408f7c11bba8601/torch/_inductor/fx_passes/post_grad.py#L903 

will pass the args in directly.

File: torch/_inductor/fx_passes/reduced_atomic_contention.py [hunk 6]
```diff
@@ -0,0 +1,582 @@
...
+    # Check minimum index size threshold
+    min_index_size = getattr(config, "partitioned_scatter_min_index_size", 4096)
+    if index_size < min_index_size:
+        log.debug(
+            "Skipping: index size %d below threshold %d", index_size, min_index_size
+        )
+        return False
+
+    # Only use if index_size is small enough and estimated contention is relevant
+    if not (index_size < (min_index_size * 8)) and contention_ratio < 4:
```
Comment (eellison) [inline torch/_inductor/fx_passes/reduced_atomic_contention.py]: configable ?

File: torch/_inductor/fx_passes/reduced_atomic_contention.py [hunk 7]
```diff
@@ -0,0 +1,582 @@
...
+    # Contention-based calculation - square root scaling
+    # Use max to ensure we never go below MIN_PARTITIONS for the base calculation
+    base_partitions = max(MIN_PARTITIONS, int(math.sqrt(contention_ratio) * 16))
+
+    # Round to power of 2 and apply limits
+    partitions = 2 ** math.ceil(math.log2(base_partitions))
+    return min(partitions, max_partitions_for_size, MAX_PARTITIONS)
+
+
+def _fit_to_memory_budget(
```
Comment (eellison) [inline torch/_inductor/fx_passes/reduced_atomic_contention.py:298]: Potentially we could use [memory estimator](torch/_inductor/fx_passes/memory_estimator.py) to estimate slack memory for high water mark.

File: torch/_inductor/fx_passes/reduced_atomic_contention.py [hunk 8]
```diff
@@ -0,0 +1,582 @@
...
+
+
+log = logging.getLogger(__name__)
+aten = torch.ops.aten
+prims = torch.ops.prims
+
+# Constants for partition configuration
+MIN_PARTITIONS = 2
+MAX_PARTITIONS = 128
+MEMORY_BUDGET_FRACTION = 0.10
```
Comment (eellison) [inline torch/_inductor/fx_passes/reduced_atomic_contention.py]: We'll need to do a second, pass over the partitions to subsequently accumulate them. What is the tradeoff of contention vs additional memory accesses ?

**Review by eellison — COMMENTED** [review]
Cool! A few questions. Would you mind trying to rewrite this with register_replacement or register_graph_pattern? I think this would be more concise for the replacement / use fewer manual apis.

My other question is the heuristics around number of partitions and memory. 

How much of this is related to memory use / risk of oom, vs tradeoff of contention vs memory access ? For the former, we could potentially restrict application based on head room for peak memory use.

**Review by eellison — COMMENTED** [review]
Cool! A few questions. Would you mind trying to rewrite this with register_replacement or register_graph_pattern? I think this would be more concise for the replacement / use fewer manual apis.

My other question is the heuristics around number of partitions and memory. 

How much of this is related to memory use / risk of oom, vs tradeoff of contention vs memory access ? For the former, we could potentially restrict application based on head room for peak memory use.

File: torch/_inductor/fx_passes/reduced_atomic_contention.py (hunk 1 above)
Comment (jataylo) [reply torch/_inductor/fx_passes/reduced_atomic_contention.py:106]: Nothing functionally, but IIUC we could choose a large number of partitions for a smaller tensor for original compilation and then later OOM out if the output tensor size increases, this was mostly for safety. Was planning to assess whether something intelligent could be done here in a follow up PR.

File: torch/_inductor/fx_passes/reduced_atomic_contention.py (hunk 2 above)
Comment (jataylo) [reply torch/_inductor/fx_passes/reduced_atomic_contention.py:109]: We could try to enable it, I was mostly trying to be selective of the use case for initial enablement. Most of the bool examples I've seen in models and UTs were small examples and wouldn't benefit from the additional kernel overhead, but this should be able to be handled by disabling the optimization when there are small number of writes.

File: torch/_inductor/fx_passes/reduced_atomic_contention.py (hunk 8 above)
Comment (jataylo) [reply torch/_inductor/fx_passes/reduced_atomic_contention.py]: It all depends on the level of atomic contention, aka the number of colliding writes between threads to locations on the same cache line. If number of collisions is high we observe a drastic slow down (see perf table in PR description).  Unfortunately we cannot estimate this at compile time as we do not number know the contents of the tensor, so we have to estimate based on the tensor shapes.

We could attempt to find the true trade off here by identifying the number of collisions required before this approach becomes beneficial but unfortunately we would have no way of applying that to the fxgraph without knowing the contents of the index tensor. We know that due to architecture difference that AMD benefits more than NV with this update, in the example provided the new approach is faster even for the best case scenario, where indices are uniformly distributed.

**Comment by jataylo** [comment]
@eellison 
UTs are passing https://hud.pytorch.org/pytorch/pytorch/pull/168073?sha=06e2520fb42dae90e2b10cf9d1168781667c3a6a besides linting with feature enabled. Now re-disabling the feature along with lint fix.

PR comments addressed with https://github.com/pytorch/pytorch/pull/168073/changes/0a56d16cf0bde72d0dc2d34ef1c9f1229ac52416#diff-ba942c6441bfa7370118e286129aef6d8644014242742c97fce1970d84a348e0

**Review by eellison — APPROVED** [review]
We can do more validation prior to enabling by default.
